From 2a166768b9bd385e517bfa47605cb6b51edafec9 Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Thu, 18 Sep 2025 23:46:06 +0530
Subject: [PATCH 3/3] ui/gtk: Re-grabbing PTR/KBD individually

Re-grabbing any of PTR and KBD is done only if both were grabbed before.
There are cases where only one of them is grabbed so it should be
checked and regrabbed separtely.

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 ui/gtk.c | 38 +++++++++++++++++---------------------
 1 file changed, 17 insertions(+), 21 deletions(-)

diff --git a/ui/gtk.c b/ui/gtk.c
index fce6fd1f99..1442507b45 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -1574,17 +1574,16 @@ static gboolean gd_window_state_event(GtkWidget *widget, GdkEvent *event,
     }
 
     if (event->window_state.new_window_state & GDK_WINDOW_STATE_FOCUSED) {
-        if (s->ptr_owner == vc && s->kbd_owner == vc) {
-            if (gd_is_grab_active(s)) {
-                s->ptr_owner = NULL;
-                s->kbd_owner = NULL;
-                gd_grab_keyboard(vc, "windows-focused");
-                gd_grab_pointer(vc, "windows-focused");
-            } else {
-                gd_ungrab_keyboard(s);
-                gd_ungrab_pointer(s);
-            }
+        if (s->ptr_owner == vc) {
+            s->ptr_owner = NULL;
+            gd_grab_pointer(vc, "windows-focused");
+        }
+
+        if (s->kbd_owner == vc) {
+            s->kbd_owner = NULL;
+            gd_grab_keyboard(vc, "windows-focused");
         }
+
     }
 
     /* WA to fullscreen window if it's forcefully un-fullscreened by
@@ -1710,17 +1709,14 @@ static void gd_window_show_on_monitor(GdkDisplay *dpy, VirtualConsole *vc,
     info.height = gdk_window_get_height(window);
     dpy_set_ui_info(vc->gfx.dcl.con, &info, false);
 
-    if (s->ptr_owner && s->kbd_owner &&
-        s->ptr_owner == vc && s->kbd_owner == vc) {
-        if (gd_is_grab_active(s)) {
-            s->ptr_owner = NULL;
-            s->kbd_owner = NULL;
-            gd_grab_keyboard(vc, "user-request-main-window");
-            gd_grab_pointer(vc, "user-request-main-window");
-        } else {
-            gd_ungrab_keyboard(s);
-            gd_ungrab_pointer(s);
-        }
+    if (s->ptr_owner == vc) {
+        s->ptr_owner = NULL;
+        gd_grab_pointer(vc, "user-request-main-window");
+    }
+
+    if (s->kbd_owner == vc) {
+        s->kbd_owner = NULL;
+        gd_grab_keyboard(vc, "user-request-main-window");
     }
 
     gd_update_cursor(vc);
-- 
2.34.1

